apps/req.c

```diff
@@ -1814,23 +1778,19 @@ int do_X509_CRL_sign(X509_CRL *x, EVP_PKEY *pkey, const EVP_MD *md,
 {
     int rv;
     EVP_MD_CTX *mctx = EVP_MD_CTX_new();
-#ifndef OPENSSL_NO_SM2
-    EVP_PKEY_CTX *pctx = NULL;
-#endif
 
     rv = do_sign_init(mctx, pkey, md, sigopts);
     if (rv > 0) {
         rv = X509_CRL_sign_ctx(x, mctx);
-#ifndef OPENSSL_NO_SM2
         /*
-         * only in SM2 case we need to free the pctx explicitly
-         * if do_sign_init() fails, no need to double free pctx
+         * With SM2, do_sign_init() attached an EVP_PKEY_CTX to the
+         * EVP_MD_CTX, and we have to free it explicitly.
          */
-        if (ec_pkey_is_sm2(pkey)) {
-            pctx = EVP_MD_CTX_pkey_ctx(mctx);
+        if (EVP_PKEY_id(pkey) == EVP_PKEY_SM2) {
+            EVP_PKEY_CTX *pctx = EVP_MD_CTX_pkey_ctx(mctx);
+
```

Comment:
I'm not fond of this piece of code repeating 3 times but have no idea how to deal with it.

Reply:
Good point, I'll refactor that.

Reply:
Done

Reply:
It's significantly better but couldn't it be done in the sm2-related code, not in general code?